### PR TITLE
[ML] Fix debug builds

### PR DESCRIPTION
**Pull request by droberts195**
Some followups to #2363 to make it possible to build with
trace logging enabled.

### CODE REVIEW

**Review by tveasey — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
This change seems to have made the following tests go into infinite loops:

```
2022-07-22 12:29:03,647381 UTC [3653] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@28 +-------------------------------------------------------------------+
2022-07-22 12:29:03,647398 UTC [3653] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@29 |  CCalendarComponentAdaptiveBucketingTest/testMinimumBucketLength  |
2022-07-22 12:29:03,647414 UTC [3653] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@30 +-------------------------------------------------------------------+
```

```
2022-07-22 12:32:08,297668 UTC [4206] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@28 +-------------------------------------------+
2022-07-22 12:32:08,297675 UTC [4206] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@29 |  CKMostCorrelatedTest/testMostCorrelated  |
2022-07-22 12:32:08,297682 UTC [4206] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@30 +-------------------------------------------+
```

```
2022-07-22 12:30:54,657204 UTC [4138] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@28 +---------------------------------------------+
2022-07-22 12:30:54,657210 UTC [4138] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@29 |  CEventRateModelTest/testCorrelatedNoTrend  |
2022-07-22 12:30:54,657215 UTC [4138] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/test/CTestObserver.cc@30 +---------------------------------------------+
2022-07-22 12:30:54,657357 UTC [4138] DEBUG /private/var/lib/jenkins/workspace/elastic+machine-learning+pull-request+macos-aarch64/lib/model/unittest/CEventRateModelTest.cc@651 Test correlation anomalies
```

Stack trace from one of these is:

```
(lldb) process attach --pid 4138
ml_test_model was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 4138 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x000000010107a678 ml_test_model`CEventRateModelTest::testCorrelatedNoTrend::test_method() [inlined] void ml::core::CScopePrintContainers::print<ml::maths::common::CBasicStatistics::COrderStatisticsHeap<ml::core::CTriple<double, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::less<ml::core::CTriple<double, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&>(this=<unavailable>, t=<unavailable>) at CContainerPrinter.h:352:21 [opt]
   349      std::ostream* stream() const { return m_Stream; }
   350      template<typename T>
   351      void print(T&& t, printer_detail::true_) {
-> 352          (*m_Stream) << t;
   353      }
   354      template<typename T>
   355      void print(T&& t, printer_detail::false_) {
```

So container printing a `COrderStatisticsHeap` now has an infinite loop.

File: include/maths/common/CBasicStatistics.h [hunk 1]
```diff
@@ -1453,6 +1454,20 @@ std::ostream& operator<<(std::ostream& o,
...
+    o << core::CScopePrintContainers{} << accumulator;
+    return o;
+}
+
+template<typename T, typename LESS>
+std::ostream& operator<<(std::ostream& o,
+                         const CBasicStatistics::COrderStatisticsHeap<T, LESS>& accumulator) {
+    o << core::CScopePrintContainers{} << accumulator;
+    return o;
+}
```
Comment (tveasey) [inline include/maths/common/CBasicStatistics.h:1469]: This actually probably doesn't work on reflection. `CScopePrintContainers` checks to see if the type is printable and finds this function, but then it implements the print using this function. So an infinite recursion. It should work if you make these instead:
```suggestion
template<typename T, std::size_t N, typename LESS>
std::ostream& operator<<(std::ostream& o,
                         const CBasicStatistics::COrderStatisticsStack<T, N, LESS>& accumulator) {
    o << core::CContainerPrinter::print(accumulator);
    return o;
}

template<typename T, typename LESS>
std::ostream& operator<<(std::ostream& o,
                         const CBasicStatistics::COrderStatisticsHeap<T, LESS>& accumulator) {
    o << core::CContainerPrinter::print(accumulator);
    return o;
}
```

File: include/model/CMetricMultivariateStatistic.h [hunk 2]
```diff
@@ -154,28 +155,21 @@ class CMetricMultivariateStatistic {
...
-
 private:
     using TStatistic2Vec = core::CSmallVector<STATISTIC, 2>;
 
 private:
     TStatistic2Vec m_Values;
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const CMetricMultivariateStatistic<STATISTIC>& statistic) {
+        o << core::CScopePrintContainers{} << statistic.m_Values;
```
Comment (tveasey) [inline include/model/CMetricMultivariateStatistic.h:166]: Same here as well

**Pull request merged**